examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,231 @@
+        }
+
+        #[ink(message)]
+        fn remove_owner(&mut self, owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_owner(&owner);
+            let len = self.owners.len() - 1;
+            let required = u32::min(len, *self.required.get());
+            ensure_requirement(len, required);
+            self.owners.swap_remove(self.owner_index(&owner));
+            self.is_owner.remove(&owner);
+            self.required.set(required);
+            self.clean_owner_confirmations(&owner);
+        }
+
+        #[ink(message)]
+        fn replace_owner(&mut self, owner: AccountId, new_owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_owner(&owner);
+            self.ensure_no_owner(&new_owner);
+            self.owners.replace(self.owner_index(&owner), || new_owner);
+            self.is_owner.remove(&owner);
+            self.is_owner.insert(new_owner, ());
+            self.clean_owner_confirmations(&owner);
+        }
+
+        #[ink(message)]
+        fn change_requirement(&mut self, requirement: u32) {
+            self.ensure_from_wallet();
+            ensure_requirement(self.owners.len(), requirement);
+            self.required.set(requirement);
+        }
+
+        #[ink(message)]
+        fn submit_transaction(&mut self, transaction: Transaction) {
+            self.ensure_from_owner();
+            let id = self.transactions.put(transaction);
+            self.confirmation_count.insert(id, 0);
+            self.internal_confirm(id);
+        }
+
+        #[ink(message)]
+        fn cancel_transaction(&mut self, id: TransactionId) {
+            self.ensure_from_wallet();
+            self.take_transaction(id);
+        }
+
+        #[ink(message)]
+        fn confirm_transaction(&mut self, id: TransactionId) {
+            self.ensure_from_owner();
+            self.ensure_transaction_exists(id);
+            self.internal_confirm(id);
+        }
+
+        #[ink(message)]
+        fn revoke_confirmation(&mut self, id: TransactionId) {
+            self.ensure_from_owner();
+            if self
+                .confirmations
+                .remove(&(id, self.env().caller()))
+                .is_some()
+            {
+                self.confirmation_count
+                    .mutate_with(&id, |count| *count -= 1);
+            }
+        }
+
+        #[ink(message)]
+        fn execute_transaction(&mut self, id: TransactionId) -> Result<(), ()> {
+            self.ensure_confirmed(id);
+            let t = self.take_transaction(id).unwrap();
+            env::call::CallParams::<MyEnv, ()>::invoke(t.callee, t.selector.into())
+                .gas_limit(t.gas_limit)
+                .transferred_value(t.transferred_value)
+                .push_arg(&CallInput(&t.input))
+                .fire()
+                .map(|_| ())
+                .map_err(|_| ())
+        }
+
+        fn internal_confirm(&mut self, id: TransactionId) {
```

Comment:
needs better name, e.g. `confirm_transaction_by_caller`

Reply:
but also I'd like to keep internal methods more descriptive, e.g. adding this `confirmer: AccountId` to the params and invoking it with `self.env().caller()` on call-site instead of using `self.env().caller()` in the implementation. but this is details.

Reply:
Done. Resolve if satisfied.